### ISSUE TITLE
Loading SimpleModules from a library

### ISSUE THREAD

**Issue by ckosmowski** [micronaut-projects/micronaut-core#2671]
Thanks for reporting an issue for Micronaut, please review the task list below before submitting the
issue. Your issue report will be closed if the issue is incomplete and the below tasks not completed.

NOTE: If you are unsure about something and the issue is more of a question a better place to ask questions is on Stack Overflow (http://stackoverflow.com/tags/micronaut) or Gitter (https://gitter.im/micronautfw/). DO NOT use the issue tracker to ask questions.

### Task List

- [X] Steps to reproduce provided
- [X] Stacktrace (if present) provided
- [ ] Example that reproduces the problem uploaded to Github (provided if needed)
- [X] Full description of the issue provided (see below)

### Steps to Reproduce

1. Implement a library containing a SimpleModule include JsonDeserializers that should integrate with Jacksons ObjectMapper for custom deserialization (In the test case within a @Factory)
2. Build the library with micronaut and annotation processors in place
3. Create a micronaut project that uses the library and that deserializes Instances of the class that the SimpleModule was created for

### Expected Behaviour

The SimpleModule from the library should be automatically available in the ObjectMapper and therefore be found for deserialization. (Everything worked as long as the same code was located in the main project itsself)

### Actual Behaviour

The SimpleModules seems not to be found, even if the @Singleton method in the @Factory is additonally annotated with @Context and not even if a package-info.java with @Configuration is provided

## StackTrace

```Caused by: io.micronaut.core.serialize.exceptions.SerializationException: Error deserializing object from JSON: Cannot construct instance of `de.rhenus.fl.dpx.commons.business.money.Monetary` (no Creators, like default construct, exist): cannot deserialize from Object value (no delegate- or property-based Creator)
 at [Source: (byte[])"{"costs":{"currency":"EUR","amount":"62.000"},"ameta":{"currency":"EUR","amount":"62.000"},"turnover":{"currency":"EUR","amount":"62.000"},"handlingCosts":{"currency":"EUR","amount":"62.000"},"accountingNumber":"1234522","accountingRole":"4","accountingSystemId":2,"operationalSystemId":1,"organisationalUnitId":35,"shipmentNumber":"202001161444492760","product":"NAT","mainOfficeNumber":"41661","month":1,"year":2020,"shipmentHash":"20200116144449272","serviceDate":"2019-03-22T14:44:48"}"; line: 1, column: 11] (through reference chain: de.rhenus.fl.dpx.shipmentfinance.operationalturnover.adapter.kafka.OperationalTurnoverMessage["costs"])
	at io.micronaut.jackson.serialize.JacksonObjectSerializer.deserialize(JacksonObjectSerializer.java:70)
	at io.micronaut.configuration.kafka.serde.JsonSerde.deserialize(JsonSerde.java:79)
	at org.apache.kafka.common.serialization.Deserializer.deserialize(Deserializer.java:60)
	at org.apache.kafka.clients.consumer.internals.Fetcher.parseRecord(Fetcher.java:1268)
	at org.apache.kafka.clients.consumer.internals.Fetcher.access$3600(Fetcher.java:124)
	at org.apache.kafka.clients.consumer.internals.Fetcher$PartitionRecords.fetchRecords(Fetcher.java:1492)
	at org.apache.kafka.clients.consumer.internals.Fetcher$PartitionRecords.access$1600(Fetcher.java:1332)
	at org.apache.kafka.clients.consumer.internals.Fetcher.fetchRecords(Fetcher.java:645)
	at org.apache.kafka.clients.consumer.internals.Fetcher.fetchedRecords(Fetcher.java:606)
	at org.apache.kafka.clients.consumer.KafkaConsumer.pollForFetches(KafkaConsumer.java:1263)
	at org.apache.kafka.clients.consumer.KafkaConsumer.poll(KafkaConsumer.java:1225)
	at org.apache.kafka.clients.consumer.KafkaConsumer.poll(KafkaConsumer.java:1201)
	at io.micronaut.configuration.kafka.processor.KafkaConsumerProcessor.lambda$process$7(KafkaConsumerProcessor.java:393)
	at io.micrometer.core.instrument.composite.CompositeTimer.record(CompositeTimer.java:79)
	at io.micrometer.core.instrument.Timer.lambda$wrap$0(Timer.java:144)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:830)
Caused by: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Cannot construct instance of `de.rhenus.fl.dpx.commons.business.money.Monetary` (no Creators, like default construct, exist): cannot deserialize from Object value (no delegate- or property-based Creator)
 at [Source: (byte[])"{"costs":{"currency":"EUR","amount":"62.000"},"ameta":{"currency":"EUR","amount":"62.000"},"turnover":{"currency":"EUR","amount":"62.000"},"handlingCosts":{"currency":"EUR","amount":"62.000"},"accountingNumber":"1234522","accountingRole":"4","accountingSystemId":2,"operationalSystemId":1,"organisationalUnitId":35,"shipmentNumber":"202001161444492760","product":"NAT","mainOfficeNumber":"41661","month":1,"year":2020,"shipmentHash":"20200116144449272","serviceDate":"2019-03-22T14:44:48"}"; line: 1, column: 11] (through reference chain: de.rhenus.fl.dpx.shipmentfinance.operationalturnover.adapter.kafka.OperationalTurnoverMessage["costs"])
	at com.fasterxml.jackson.databind.exc.InvalidDefinitionException.from(InvalidDefinitionException.java:67)
	at com.fasterxml.jackson.databind.DeserializationContext.reportBadDefinition(DeserializationContext.java:1589)
	at com.fasterxml.jackson.databind.DeserializationContext.handleMissingInstantiator(DeserializationContext.java:1055)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1297)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:326)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:159)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:288)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:151)
	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4202)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3266)
	at io.micronaut.jackson.serialize.JacksonObjectSerializer.deserialize(JacksonObjectSerializer.java:68)
	... 19 common frames omitted```

### Environment Information

- **Operating System**: Windows 10
- **Micronaut Version:** 1.3.0.M2
- **JDK Version:** OpenJDK 8

### Example Application

- Provided if needed to track down the problem

**Comment by jameskleeh** [comment]
Please upload an example that reproduces the issue

**Comment by ckosmowski** [comment]
This will be really hard and i will probably rather put the code back to the main project, because i would have to publish the library project to a public repository.

Could you verify that the scenario where you have beans in a library annotated with @Singleton processed with the annotation processor SHOULD work in general?

**Comment by jameskleeh** [comment]
@ckosmowski Yes that should work

**Comment by ckosmowski** [comment]
After some debugging, it seems the SimpleModule Bean is available and is injected when referenced by another class but unfortunately is not added to the ObjectMapper.

**Comment by ckosmowski** [comment]
It seems to me, that the reason that the SimpleModule beans that are generally available are not found ist that at the point in time where ObjectMapper factory is executed and instatiated the beans from the library are not yet present.

Is there a chance, that this could have something to to with:
https://github.com/micronaut-projects/micronaut-core/blob/7a25c5e4bab692412f534d9aa00d07420a933907/runtime/src/main/java/io/micronaut/jackson/ObjectMapperFactory.java#L43

Unfortunately there is nothing injected here:
https://github.com/micronaut-projects/micronaut-core/blob/7a25c5e4bab692412f534d9aa00d07420a933907/runtime/src/main/java/io/micronaut/jackson/ObjectMapperFactory.java#L53

or here:

https://github.com/micronaut-projects/micronaut-core/blob/7a25c5e4bab692412f534d9aa00d07420a933907/runtime/src/main/java/io/micronaut/jackson/ObjectMapperFactory.java#L59

allthough the beans are available in other beans than the ObjectMapperFactory

**Comment by jameskleeh** [comment]
@ckosmowski Does your module bean have `@BootstrapContextCompatible` ? If not, I would add that and try again

**Comment by ckosmowski** [comment]
I already tried that, but it still doesn't work. 
I found a workaround now. I implemented a BeanCreatedListener inside the library that listens for the ObjectMapper and than i add the module manually (I found this in the documentation). This works properly.

But what is confusing about this, is that Modules inside the core projects itsself are properly found while Modules in the library are not found at least not at the moment when the ObjectMapperFactory is initialized.
